source/common/http/header_map_impl.cc

```diff
@@ -320,7 +326,14 @@ void HeaderMapImpl::insertByKey(HeaderString&& key, HeaderString&& value) {
 }
 
 void HeaderMapImpl::addViaMove(HeaderString&& key, HeaderString&& value) {
-  insertByKey(std::move(key), std::move(value));
+  // If this is an inline header, we can't addViaMove, because we'll overwrite
+  // the existing value.
+  auto* entry = getExistingInline(key.c_str());
+  if (entry != nullptr) {
+    appendToHeader(entry->value(), value.c_str());
```

Comment:
@alyssawilk I'm pretty sure the bug is here. The `addViaMove` function expects both parameters to be moved/cleared, which doesn't happen in this path. I think in this case we need to explicitly clear `value`. See https://github.com/envoyproxy/envoy/blob/28cea916a2a23c25d9175373c0413610586f5a16/source/common/http/http1/codec_impl.cc#L301. We should add more asserts at the end of this function such as `ASSERT(value.empty());`

cc @junr03 @danielhochman

Reply:
I can explicitly clear, but ASSERT(value.empty()) doesn't pass with the existing insertByKey() function.  We explicitly test in HeaderStringTest.All that when we std::move(string) the "original" string retains a pointer at the underlying buffer and the size is not cleared.  

I've added an integration test instead, which trips the ASSERT you called out, and still fails with it commented out, due to "eep" not being found.
https://github.com/envoyproxy/envoy/pull/3597